// ion-js/dist/commonjs/es6/dom/Integer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Integer = void 0;
const jsbi_1 = __importDefault(require("jsbi"));
const Ion_1 = require("../Ion");
const FromJsConstructor_1 = require("./FromJsConstructor");
const Value_1 = require("./Value");
const _jsbiConstructor = jsbi_1.default;
const _fromJsConstructor = new FromJsConstructor_1.FromJsConstructorBuilder()
    .withPrimitives(FromJsConstructor_1.Primitives.Number)
    .withClassesToUnbox(Number)
    .withClasses(_jsbiConstructor)
    .build();
class Integer extends Value_1.Value(Number, Ion_1.IonTypes.INT, _fromJsConstructor) {
    constructor(value, annotations = []) {
        if (typeof value === "number") {
            super(value);
            this._numberValue = value;
            this._bigIntValue = null;
        }
        else {
            const numberValue = jsbi_1.default.toNumber(value);
            super(numberValue);
            this._bigIntValue = value;
            this._numberValue = numberValue;
        }
        this._setAnnotations(annotations);
    }
    bigIntValue() {
        if (this._bigIntValue === null) {
            this._bigIntValue = jsbi_1.default.BigInt(this.numberValue());
        }
        return this._bigIntValue;
    }
    numberValue() {
        return this._numberValue;
    }
    toString() {
        if (this._bigIntValue === null) {
            return this._numberValue.toString();
        }
        return this._bigIntValue.toString();
    }
    valueOf() {
        return this.numberValue();
    }
    writeTo(writer) {
        writer.setAnnotations(this.getAnnotations());
        if (this._bigIntValue === null) {
            writer.writeInt(this.numberValue());
        }
        else {
            writer.writeInt(this._bigIntValue);
        }
    }
    _valueEquals(other, options = {
        epsilon: null,
        ignoreAnnotations: false,
        ignoreTimestampPrecision: false,
        onlyCompareIon: true,
    }) {
        let isSupportedType = false;
        let valueToCompare = null;
        if (other instanceof Integer) {
            isSupportedType = true;
            if (this._bigIntValue == null && other._bigIntValue == null) {
                valueToCompare = other.numberValue();
            }
            else {
                valueToCompare = other.bigIntValue();
            }
        }
        else if (!options.onlyCompareIon) {
            if (other instanceof Number || typeof other === "number") {
                isSupportedType = true;
                if (this.bigIntValue == null) {
                    valueToCompare = other.valueOf();
                }
                else {
                    valueToCompare = jsbi_1.default.BigInt(other.valueOf());
                }
            }
            else if (other instanceof jsbi_1.default) {
                isSupportedType = true;
                valueToCompare = other;
            }
        }
        if (!isSupportedType) {
            return false;
        }
        if (valueToCompare instanceof jsbi_1.default) {
            return jsbi_1.default.equal(this.bigIntValue(), valueToCompare);
        }
        return this.numberValue() == valueToCompare;
    }
}
exports.Integer = Integer;
